### PR TITLE
Support streaming uploads and downloads

### PR DESCRIPTION
**Pull request by ahjmorton**
This PR adds an additional `stream` API which allows uploads and downloads to be driven by IO like objects.

To do this it changes the way adapters work to think in terms of IO like objects for upload / download. It also maintains the existing String based interface by making it a special case of IO like operations e.g. we wrap Strings in `StringIO` to keep the API consistent.

This PR also steals a ton of content from #63 including whole commits / approach

### CODE REVIEW

File: lib/bucket_store/key_storage.rb [hunk 1]
```diff
@@ -60,25 +60,27 @@ def download
     # @param [String] content The content to upload
     # @return [String] The final `key` where the content has been uploaded
     # @example Upload a file
-    #   BucketStore.for("inmemory://bucket/file.xml").upload("hello world")
+    #   BucketStore.for("inmemory://bucket/file.xml").upload!("hello world")
     def upload!(content)
```
Comment (ahjmorton) [inline lib/bucket_store/key_storage.rb:64]: An alternative approach here is assuming the content is a file then having a `if content.instance_of?(String)` to wrap Strings in StringIO. For the sake of backwards compatibility / explictness I decided against that approach but happy to be convinced that it's a better one

Comment (ivgiuliani) [reply lib/bucket_store/key_storage.rb:64]: That would be backwards compatible, as it doesn't require any changes to existing codebases so the main thing to understand is whether the unified API is a better choice compared to having explicit some separation.

Wonder if the this is a case better served by streaming uploads (see #63): if you use the non-streaming version, then only strings can be uploaded, but file-like objects are handled by `.stream`. This will be the case implicitly here, since under the hood s3/gcs will stream file reads. Though I guess it opens a question of whether the interfaced proposed in #63 is the right one... WDYT? Happy to go either way, though I'd slightly prefer a single `.upload!` method to reduce cognitive load when using the gem

Comment (ahjmorton) [reply lib/bucket_store/key_storage.rb:64]: So I actually missed that #63 added support for streaming uploads as the PR title only mentioned downloads. I see now there is an upload version with `NotImplemented` on it.

> Though I guess it opens a question of whether the interfaced proposed in https://github.com/gocardless/bucket-store/pull/63 is the right one... WDYT? Happy to go either way, though I'd slightly prefer a single .upload! method to reduce cognitive load when using the gem

I will say that from a "least surprise" perspective having a single `upload!` method that accepts a string or a file like object and handles the difference behind the scenes seems to be a common approach within Ruby e.g. the S3 SDK does this for file uploads vs exposing a separate `put_object_from_file` method / explicit `.stream` version. Naturally there are counter examples to this but want to say the spirit of "do what I mean" seems to prevail within other libraries.

In terms of API consistency I'm also wondering if exposing streaming downoads via an IO like API might fit better with existing libraries e.g. https://github.com/soulcutter/saxerator accepts a file like object for XML parsing.

WDYR? Happy to have a call on this

File: lib/bucket_store/key_storage.rb [hunk 2]
```diff
@@ -40,45 +124,31 @@ def filename
...
-      result = adapter.download(bucket: bucket, key: key)
-
-      BucketStore.logger.info(event: "key_storage.download_finished",
-                              duration: BucketStore::Timing.monotonic_now - start)
-
-      result
+      buffer = StringIO.new
+      stream.download(file: buffer).
+        except(:file).
+        merge(content: buffer.string)
```
Comment (ahjmorton) [inline lib/bucket_store/key_storage.rb:130]: Not that happy that we're creating a couple of hashes here. Imagine memory usage of this is marginal but might end up with more GC pressure

Comment (ahjmorton) [reply lib/bucket_store/key_storage.rb]: well, fun fact: `except` isn't available in Ruby 2.7 which this needs to support so going with a more memory safe option of altering the hash inline

File: lib/bucket_store/key_storage.rb [hunk 3]
```diff
@@ -15,6 +15,88 @@ class KeyStorage
       disk: Disk,
     }.freeze
 
+    # Defines a streaming interface for download and upload operations.
+    #
+    # Note that individual adapters may require additional configuration for the correct
+    # behavior of the streaming interface.
+    class KeyStreamer
+      attr_reader :bucket, :key, :adapter, :adapter_type
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:23]: `adapter` needs to be private to be consistent with the outer class

File: lib/bucket_store/key_storage.rb [hunk 4]
```diff
@@ -15,6 +15,88 @@ class KeyStorage
...
+
+      def initialize(adapter:, adapter_type:, bucket:, key:)
+        @adapter = adapter
+        @adapter_type = adapter_type
+        @bucket = bucket
+        @key = key
+      end
+
+      # Streams the content of the reference key into a File like object
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:32]:
```suggestion
      # Streams the content of the reference key into a file-like object
```

File: lib/bucket_store/key_storage.rb [hunk 5]
```diff
@@ -15,6 +15,88 @@ class KeyStorage
...
+      #
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:33]:
```suggestion
      # @param [IO] file a writeable IO instance, or a file-like object such as `StringIO`
```

File: lib/bucket_store/key_storage.rb [hunk 6]
```diff
@@ -15,6 +15,88 @@ class KeyStorage
...
+
+      # Performs a streaming upload to the backing object store
+      #
+      # @return the generated key for the new object
+      #
+      # @see KeyStorage#upload!
+      # @example Upload a key
+      #   buffer = StringIO.new("Imagine I'm a 2GB file")
+      #   BucketStore.for("inmemory://bucket/file.xml").stream.upload!(file: buffer)
+      def upload!(file:)
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:69]: Needs a similar `@param` to downloads

File: lib/bucket_store/key_storage.rb [hunk 7]
```diff
@@ -40,45 +122,32 @@ def filename
     # @example Download a key
     #   BucketStore.for("inmemory://bucket/file.xml").download
     def download
-      raise ArgumentError, "Key cannot be empty" if key.empty?
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:43]: Don't we still need this?

File: lib/bucket_store/key_storage.rb [hunk 8]
```diff
@@ -40,45 +122,32 @@ def filename
...
-
-      BucketStore.logger.info(event: "key_storage.download_finished",
-                              duration: BucketStore::Timing.monotonic_now - start)
-
-      result
+      buffer = StringIO.new
+      stream.download(file: buffer).tap do |result|
+        result.delete(:file)
+        result[:content] = buffer.string
+      end
```
Comment (ivgiuliani) [inline lib/bucket_store/key_storage.rb:129]: I think we're changing the logging structure with this approach as everything is now logged under `key_storage.stream.*`. I can live with this but seems like a breaking change? Perhaps we can just remove the `.stream` from `KeyStreamer` though

File: spec/bucket_store/disk_spec.rb [hunk 9]
```diff
@@ -112,11 +111,11 @@
 
     context "when the bucket has some keys in it" do
       before do
-        instance.upload!(bucket: bucket, key: "2019-01/hello1", content: "world")
-        instance.upload!(bucket: bucket, key: "2019-01/hello2", content: "world")
-        instance.upload!(bucket: bucket, key: "2019-01/hello3", content: "world")
-        instance.upload!(bucket: bucket, key: "2019-02/hello", content: "world")
-        instance.upload!(bucket: bucket, key: "2019-03/hello", content: "world")
+        instance.upload!(bucket: bucket, key: "2019-01/hello1", file: file)
+        instance.upload!(bucket: bucket, key: "2019-01/hello2", file: file)
```
Comment (ivgiuliani) [inline spec/bucket_store/disk_spec.rb:115]: How does this work? I would have guessed that we need a `.rewind` here? 🤔

File: spec/bucket_store_integration_spec.rb [hunk 10]
```diff
@@ -24,55 +26,129 @@
...
+            expect(described_class.for(base_bucket_uri.to_s).list.to_a.size).to eq(0)
+          end
+        end
+
+        # Ruby files have a binary / textual mode.
+        # This means if you try and stick a byte array
+        # into a file it'll silently encode the bytes to
+        # whatever encoding is setup in the file.
+        # To validate byte arrays work we create binmode
+        # versions of input / output and the buffer.
```
Comment (ivgiuliani) [inline spec/bucket_store_integration_spec.rb:86]: We can go a bit wider here, would be slightly easier to read

File: spec/bucket_store_integration_spec.rb [hunk 11]
```diff
@@ -24,55 +26,129 @@
...
+
+            # Now let's try writing to another temp file
+            described_class.for("#{base_bucket_uri}/prefix/#{filename}").
+              stream.
+              download(file: file_to_write)
+
+            file_to_write.rewind
+            expect(file_to_write.read).to eq(input_bytes)
+
+            # Santity check the list API behaves / interplays well with the streaming API
```
Comment (ivgiuliani) [inline spec/bucket_store_integration_spec.rb:70]:
```suggestion
            # Sanity check the list API behaves / interplays well with the streaming API
```

File: spec/bucket_store_integration_spec.rb [hunk 12]
```diff
@@ -24,55 +26,129 @@
...
-      end
-      all_files.each do |content, key|
-        expect(described_class.for(key).download[:content]).to eq(content)
+      context "when handling streaming uploads / downloads" do
+        after do
+          file_to_read.close
+          file_to_write.close
+        end
+
+        shared_examples "consistently handles different file modes" do
```
Comment (ivgiuliani) [inline spec/bucket_store_integration_spec.rb:42]: I'm a bit wary of `shared_examples` within another shared example as it's starting to be hard to follow what's going on. Wonder if we can simplify these tests a bit, and split them out somehow? With the delete-all in the `before` block, I reckon we can isolate things a bit more now instead of having a single massive test

**Review by ivgiuliani — COMMENTED** [review]
Some other minor comments:
- we need some instructions in the README for the new methods
- we should update the changelog

**Comment by ahjmorton** [comment]
> Some other minor comments:
> 
> * we need some instructions in the README for the new methods
> * we should update the changelog

Added both now @ivgiuliani

File: lib/bucket_store/key_storage.rb (hunk 6 above)
Comment (ahjmorton) [reply lib/bucket_store/key_storage.rb:69]: done

File: lib/bucket_store/key_storage.rb (hunk 8 above)
Comment (ahjmorton) [reply lib/bucket_store/key_storage.rb:131]: So I defaulted back to not changing the log line e.g. if you're using `streaming` or regular you'll get the same logs. Couldn't figure out a good ROI story for separating them

File: lib/bucket_store/key_storage.rb (hunk 7 above)
Comment (ahjmorton) [reply lib/bucket_store/key_storage.rb:43]: Nope: when we call `stream` it does this check for both uploads / downloads

File: spec/bucket_store/disk_spec.rb (hunk 9 above)
Comment (ahjmorton) [reply spec/bucket_store/disk_spec.rb:115]: 🤔 so I think this is working because it's creating a bunch of empty files within the backing store and the spec itself isn't verifying contents: it's just counting numbers of files. 

Can look to fix it up

Comment (ahjmorton) [reply spec/bucket_store/disk_spec.rb]: Fixed this

File: spec/bucket_store_integration_spec.rb (hunk 10 above)
Comment (ahjmorton) [reply spec/bucket_store_integration_spec.rb]: Updated

File: spec/bucket_store_integration_spec.rb (hunk 12 above)
Comment (ahjmorton) [reply spec/bucket_store_integration_spec.rb]: Re-worked it a bit and made binary files somewhat of a special case e.g. listing files / deleting shouldn't care if the files stored in GCS are binary or textual

File: lib/bucket_store/in_memory.rb [hunk 13]
```diff
@@ -24,21 +24,17 @@ def reset!
...
     end
 
-    def download(bucket:, key:)
-      {
-        bucket: bucket,
-        key: key,
-        content: @buckets[bucket].fetch(key),
-      }
+    def download(bucket:, key:, file:)
+      file.write(@buckets[bucket].fetch(key))
```
Comment (ivgiuliani) [inline lib/bucket_store/in_memory.rb:37]: Think we're missing a `.rewind` here, as otherwise we can't read the buffer back straight away which is what gcs/s3 allow (similar thing for the disk adapter)
```ruby
[1] pry(main)> require "bucket_store"
=> true
[2] pry(main)> BucketStore.for("inmemory://bucket/test.txt").upload!("hello")
I, [2023-04-20T09:57:16.487789 #78930]  INFO -- : {:event=>"key_storage.upload_started", :bucket=>"bucket", :key=>"test.txt", :adapter_type=>:inmemory}
I, [2023-04-20T09:57:16.487939 #78930]  INFO -- : {:event=>"key_storage.upload_finished", :duration=>1.4999997802078724e-05, :bucket=>"bucket", :key=>"test.txt", :adapter_type=>:inmemory}
=> "inmemory://bucket/test.txt"
[3] pry(main)> s = StringIO.new
=> #<StringIO:0x000000011859e2f0>
[4] pry(main)> BucketStore.for("inmemory://bucket/test.txt").stream.download(file: s)
I, [2023-04-20T09:57:26.061935 #78930]  INFO -- : {:event=>"key_storage.download_started"}
I, [2023-04-20T09:57:26.061979 #78930]  INFO -- : {:event=>"key_storage.download_finished", :duration=>8.000002708286047e-06}
=> {:bucket=>"bucket", :key=>"test.txt", :file=>#<StringIO:0x000000011859e2f0>}
[5] pry(main)> s.read
=> ""
[6] pry(main)> s.rewind
=> 0
[7] pry(main)> s.read
=> "hello"
```

File: spec/bucket_store_integration_spec.rb [hunk 14]
```diff
@@ -24,55 +26,162 @@
...
+          end
+          let(:output_in_memory_buffer) do
+            StringIO.new.binmode
+          end
+
+          it "files are downloadable from the store" do
+            described_class.for("#{base_bucket_uri}/#{temp_filename}").
+              stream.
+              download(file: output_temp_file)
+            output_temp_file.rewind
```
Comment (ivgiuliani) [inline spec/bucket_store_integration_spec.rb:125]: This test passes both before and after adding the different `.rewind` calls in the download methods, since its doing the rewind itself. We need to delete this `.rewind` call (and all the other ones on the output buffers in these new specs) as otherwise we are not testing the correct behaviour.

File: lib/bucket_store/in_memory.rb [hunk 15]
```diff
@@ -24,21 +24,20 @@ def reset!
...
-    def download(bucket:, key:)
-      {
-        bucket: bucket,
-        key: key,
-        content: @buckets[bucket].fetch(key),
-      }
+    def download(bucket:, key:, file:)
+      file.tap do |f|
+        f.write(@buckets[bucket].fetch(key))
+        f.rewind
```
Comment (ivgiuliani) [inline lib/bucket_store/in_memory.rb:39]: We need to do something similar in the disk adapter as well. I was also testing GCS locally and I get inconsistent behaviour now via the emulator (as in - sometimes I get the right output, sometimes don't). Might be worth popping a `.rewind` in the gcs adapter too while we're here to make sure we always get a readable stream back.

**Review by ivgiuliani — CHANGES_REQUESTED** [review]
Left a few more comments!

File: spec/bucket_store/key_storage_spec.rb [hunk 16]
```diff
@@ -117,6 +117,90 @@ def build_for(key)
...
+    context "when we try to upload a bucket" do
+      it "raises an error" do
+        expect { build_for("inmemory://bucket").stream }.
+          to raise_error(ArgumentError, /key cannot be empty/i)
+      end
+    end
+
+    describe "#download" do
+      let(:input_file_1) { StringIO.new("content1") }
+      let(:input_file_2) { StringIO.new("content") }
```
[the remark this answers is not in the document]
Comment (piiraa) [reply spec/bucket_store/key_storage_spec.rb:137]: Removed with next commit

File: lib/bucket_store/in_memory.rb (hunk 15 above)
Comment (piiraa) [reply lib/bucket_store/in_memory.rb:42]: ✔️

**Review by ivgiuliani — CHANGES_REQUESTED** [review]
Couple of things:

- the `.rewind`s should only be necessary for the download part. In the tests we should retain the `.rewind` in the test logic on the _input_ buffers, but _not_ on the output buffers
- not quite sure why tests are failing -- looks like the build has been cancelled? But we need a green tick before merging

File: spec/bucket_store_integration_spec.rb [hunk 17]
```diff
@@ -24,55 +26,162 @@
...
+
+          described_class.for("#{base_bucket_uri}/#{buffer_filename}").
+            stream.
+            upload!(file: input_in_memory_buffer)
+          input_in_memory_buffer.rewind
+        end
+
+        after do
+          input_temp_file.close
+          output_temp_file.close
```
[the remark this answers is not in the document]
Comment (thekeshavgoel) [reply spec/bucket_store_integration_spec.rb:70]: @piiraa  - Could you add unlinking the temp files here, it's a good practice to do so - https://ruby-doc.org/stdlib-2.5.0/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Unlink+after+creation

File: lib/bucket_store/in_memory.rb (hunk 13 above)
Comment (thekeshavgoel) [reply lib/bucket_store/in_memory.rb]: @piiraa  - think this still needs doing as well?

File: spec/bucket_store/key_storage_spec.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (thekeshavgoel) [reply spec/bucket_store/key_storage_spec.rb]: still here

File: spec/bucket_store_integration_spec.rb [hunk 18]
```diff
@@ -4,6 +4,10 @@
 
 require "aws-sdk-s3"
 
+require "tempfile"
+
+require "pry-byebug"
```
Comment (thekeshavgoel) [inline spec/bucket_store_integration_spec.rb:9]:
```suggestion
```

✂️

File: spec/bucket_store_integration_spec.rb [hunk 19]
```diff
@@ -24,55 +28,167 @@
...
+            input_temp_file.close
+          else
+            input_temp_file.unlink
+          end
+
+          if output_temp_file.is_a?(File)
+            output_temp_file.close
+          else
+            output_temp_file.unlink
+          end
```
Comment (thekeshavgoel) [inline spec/bucket_store_integration_spec.rb:81]: Should be just ? :
```
output_temp_file.close
output_temp_file.unlink
```

Comment (piiraa) [reply spec/bucket_store_integration_spec.rb]: This was the issue cause
```ruby
Tempfile.new.binmode.class
=> File
# vs.
Tempfile.new(binmode: true).class
=> Tempfile
```

Comment (thekeshavgoel) [reply spec/bucket_store_integration_spec.rb]: gotcha, thanks!

**Review by thekeshavgoel — APPROVED** [review]
LGTM
